FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+4. Transparent gas accounting and automatic execution halting.
+5. (Future) Dynamic linkage of popular WASM modules "blessed" by the community
+   through some governance process, in order to reduce actor bytecode size (e.g.
+   FVM SDKs, IPLD codecs, etc).
+
+Because this contract may change over time, user-deployed actors may specify an
+IC version at [`InitActor#InstallActor` time](#actor-deployment). The strategy
+for deprecating IC versions over time has not been yet determined.
+
+#### Syscalls
+
+Syscalls are functions made available to the actor as WASM imports, allowing it
+to interact with the outer environment, and to call complex operations that are
+more performant to implement in native land (e.g. cryptographic primitives).
+
+#### Kernel
+
+The Kernel is an object created by the Call Manager and injected into the
+Invocation Container. It processes syscalls, and stores state concerning the
+invocation.
+
+#### Externs
+
+Similar to syscalls being functions provided to the actor by the FVM, Externs
+are functions provided by the node to the FVM. Depending on the programming
+languages at play, Externs may need to traverse an FFI boundary.
+
+### Actors
+
+The term _Actor_ is a reference to the [actor model], a concurrent computation
+paradigm that inspires Filecoin's programming and scalability philosophies.
+
+#### Actor types
+
+We distinguish three types of actors:
+
+1. Built-in actors.
+2. Native actors.
+3. Foreign actors.
+
+![Actor execution](./fip-draft_foundational_fvm/overview.png)
+
+##### Built-in actors
+
+Built-in actors are protocol-defined actors whose WASM bytecode ships with the
+Filecoin client. The recommended approach is to embed a CAR bundle with all the
+relevant bytecode, and seeding the blockstore with it on node start.
+
+**Canonical codebase**
+
+With the definition of a universal, portable execution bytecode format, all
+Filecoin clients can rely on a single codebase for built-in actors. This speeds
+up protocol development, shortens protocol upgrade timelines, and reduces
+coordination overhead.
+
+This FIP proposes to promote the FVM built-in actors codebase (forked from the
+Forest client), to the status of **_canonical_ built-in actors**. Clients should
+adopt this codebase, but they may choose to pursue their own.
+
+##### Native actors
+
+Native actors are user-defined actors targeting the native FVM runtime.
+
+Users can _technically_ write native actors in any programming that compiles to
+WASM. However, language-specific overheads (e.g. runtime, garbage collection,
+stdlibs, etc.) may lead to oversized WASM bytecode and execution overheads,
+translating to higher gas costs.
+
+Furthermore, there may be hard limits enforced on the sizing of bytecode
+deployables in the future.
+
+There's a [reference SDK] written in
+Rust to accompany this FIP. It is used by the canonical built-in actors.
+
+We find that Rust is a reasonable choice to writing native actors due to the
```

Comment:
Well... we're currently using std. Although I guess we'd like to remove it.

Reply:
Indeed.